backend/lib/richard_burton_web/controllers/publication_controller.ex

```diff
@@ -25,10 +25,27 @@ defmodule RichardBurtonWeb.PublicationController do
     conn |> put_status(status) |> json(response_body)
   end
 
-  def parse(conn, %{"csv" => %Plug.Upload{path: path}}) do
+  def validate(conn, %{"csv" => %Plug.Upload{path: path}}) do
     try do
       publications = Publication.from_csv!(path)
-      conn |> put_status(:ok) |> json(publications)
+
+      publications_with_errors =
+        publications
+        |> Enum.map(&Publication.validate/1)
+        |> Enum.zip(publications)
+        |> Enum.map(fn tuple ->
+          case tuple do
+            {{:ok}, publication} ->
+              %{publication: publication, errors: nil}
+
+            {{:error, errors}, publication} ->
+              %{publication: publication, errors: errors}
+          end
```

Comment:
I think you could prevent from iterating three times over the publication list by doing something like this:

```elixir
publications_with_errors = Enum.map(publications, fn publication ->
  case Publication.validate(publication) do
    {:ok} ->  %{publication: publication, errors: nil}
    {:error, errors} -> %{publication: publication, errors: errors}
  end
end
   
```